examples/erc1155/lib.rs

```diff
@@ -0,0 +1,738 @@
+/// if a contract does not support this interface any transactions originating from an ERC-1155
+/// compliant contract which attempt to transfer tokens directly to the contract's address must be
+/// reverted.
+#[ink::trait_definition]
+pub trait Erc1155TokenReceiver {
+    /// Handle the receipt of a single ERC-1155 token.
+    ///
+    /// This should be called by a compliant ERC-1155 contract if the intended recipient is a smart
+    /// contract.
+    ///
+    /// If the smart contract implementing this interface accepts token transfers then it must
+    /// return `ON_ERC_1155_RECEIVED_SELECTOR` from this function. To reject a transfer it must revert.
+    ///
+    /// Any callers must revert if they receive anything other than `ON_ERC_1155_RECEIVED_SELECTOR` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_id: TokenId,
+        value: Balance,
+        data: Vec<u8>,
+    ) -> Vec<u8>;
+
+    /// Handle the receipt of multiple ERC-1155 tokens.
+    ///
+    /// This should be called by a compliant ERC-1155 contract if the intended recipient is a smart
+    /// contract.
+    ///
+    /// If the smart contract implementing this interface accepts token transfers then it must
+    /// return `BATCH_ON_ERC_1155_RECEIVED_SELECTOR` from this function. To reject a transfer it must revert.
+    ///
+    /// Any callers must revert if they receive anything other than `BATCH_ON_ERC_1155_RECEIVED_SELECTOR` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_batch_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_ids: Vec<TokenId>,
+        values: Vec<Balance>,
+        data: Vec<u8>,
+    );
+}
+
+#[ink::contract]
+mod erc1155 {
+    use super::*;
+
+    #[allow(unused_imports)]
+    use ink_env::call::{
+        build_call,
+        utils::ReturnType,
+        ExecutionInput,
+        Selector,
+    };
+
+    use ink_prelude::collections::BTreeMap;
+    use ink_storage::traits::{
+        PackedLayout,
+        SpreadLayout,
+    };
+
+    /// Indicate that a token transfer has occured.
+    ///
+    /// This must be emitted even if a zero value transfer occurs.
+    #[ink(event)]
+    pub struct TransferSingle {
+        #[ink(topic)]
+        operator: Option<AccountId>,
+        #[ink(topic)]
+        from: Option<AccountId>,
+        #[ink(topic)]
+        to: Option<AccountId>,
+        token_id: TokenId,
+        value: Balance,
+    }
+
+    /// Indicate that an approval event has happened.
+    #[ink(event)]
+    pub struct ApprovalForAll {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        operator: AccountId,
+        approved: bool,
+    }
+
+    /// Indicate that a token's URI has been updated.
+    #[ink(event)]
+    pub struct Uri {
+        value: ink_prelude::string::String,
```

Comment:
Why do you prefer using an utf8 interpreted byte sequence instead of a `Vec<u8>` for example?

Reply:
In the spec they use a Solidity `string`, which is a ["Dynamically-sized UTF-8-encoded string"](https://docs.soliditylang.org/en/develop/types.html#dynamically-sized-byte-array).